14 2secondag

2secondag opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[수 찾기](https://www.acmicpc.net/problem/1920)
## ✔️ 소요된 시간
20분

## ✨ 수도 코드
#### 문제 설명
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/1e8555ee-d2a2-42f9-8152-1178219a7f56)

#### 실패 코드 
```python
import sys

N = int(sys.stdin.readline())
N_list = list(map(int,sys.stdin.readline().split()))

M = int (sys.stdin.readline())
M_list = list(map(int,sys.stdin.readline().split()))

index = 0

for i in range(M):
  for j in range(N):
    if M_list[i] == N_list[j]:
      index = 1
      break
  print(index)
  index = 0
```
의 결과는 
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/887edd02-f57e-46ae-a798-c94de6f75b00)

이중 for문에서 시간이 많이 걸렸나? 다른 문제가 있나 싶어서 질문게시판에 들어가보니 list대신 set으로 정의하면 풀린다는 글을 발견했다. 

둘 다 수정해서! 

#### 수정한 부분
```python
N_list = map(int,sys.stdin.readline().split())
N_set = set(N_list)
```
+ `N의 배열(N_set)` 의 경우에는 순서에 상관없이 M 안에 존재하는지만 확인하면 되기 때문에 set으로 바꿔도 문제가 없다. 
 
 + `M의 배열(M_list)`의 경우에는 순서가 중요하기 때문에 list도 남겨두었다. 

```python
for i in M_list:
  if i in N_set:
    print(1)
  else:
    print(0)
```
+ 이중 for문 대신에 `in`을 사용하여 존재하는지 확인한다 

## 📚 새롭게 알게된 내용
https://blog.naver.com/jazzlubu/223354417057

#### set
집합과 같은 자료형
`s1 = set({1,2,3})` 과 `s2 = set([1,2,3])`은 표현의 차이는 있지만 같은 집합이다. 

+ **중복을 허용하지 않**으며 **순서가 없다**는 특징
 + 순서가 없기 때문에 인덱싱을 사용할 수 없음

#### 그러면 왜 liste대신 set을 사용해야하는지! 
리스트에서의 in 은 O(N) 이고 세트에서의 in 은 O(1) 의 시간복잡도를 가지기 때문!

2secondag committed:
2024.02.22

2secondag committed:
Update README.md

9kyo-hwang commented:
```py
N, A, M = int(input()), set(input().split()), int(input())
print('\n'.join('1' if num in A else '0' for num in input().split()))
```
짜잔 두 줄 마술

bomik0221 reviewed (approved):
저도 초은님이랑 나름 유사한 방식으로 풀어봤는데(find 기능을 사용), 계속 시간 초과가 나서 보니 이분 탐색으로 풀어야 하는 문제인 것 같네요..!! 그런데...  파이썬은 깔쌈하게 무시하고 편리함으로 승부를.. 오늘도 c++ 1패... ㅠㅠㅠ

2secondag commented:
> ```python
> N, A, M = int(input()), set(input().split()), int(input())
> print('\n'.join('1' if num in A else '0' for num in input().split()))
> ```
> 
> 짜잔 두 줄 마술

헐.

miniron-v reviewed (commented):
C++에서 set은 **중복을 허용하지 않으며, 순서가 있는** 자료구조로 쓰이죠. (균형 이진 트리 기반)

특정 수의 존재 여부를 $O(log N)$만에 찾을 수 있어 매우 효과적입니다.
